pkg/controllers/status/common.go

```diff
@@ -105,43 +106,53 @@ func updateResourceStatus(
 ) error {
-	gvr, err := restmapper.GetGroupVersionResource(restMapper, schema.FromAPIVersionAndKind(resource.GetAPIVersion(), resource.GetKind()))
+	gvr, err := restmapper.GetGroupVersionResource(restMapper, schema.FromAPIVersionAndKind(objRef.APIVersion, objRef.Kind))
 	if err != nil {
-		klog.Errorf("Failed to get GVR from GVK(%s/%s), Error: %v", resource.GetAPIVersion(), resource.GetKind(), err)
+		klog.Errorf("Failed to get GVR from GVK(%s/%s), Error: %v", objRef.APIVersion, objRef.Kind, err)
 		return err
 	}
 
-	if !interpreter.HookEnabled(resource.GroupVersionKind(), configv1alpha1.InterpreterOperationAggregateStatus) {
+	gvk := schema.GroupVersionKind{Group: gvr.Group, Version: gvr.Version, Kind: objRef.Kind}
+	if !interpreter.HookEnabled(gvk, configv1alpha1.InterpreterOperationAggregateStatus) {
 		return nil
 	}
+
+	// get resource template directly from apiserver, rather than informer cache, detail refers to:
+	// https://github.com/karmada-io/karmada/issues/5285
+	resource, err := dynamicClient.Resource(gvr).Namespace(objRef.Namespace).Get(ctx, objRef.Name, metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// It might happen when the resource template has been removed but the garbage collector hasn't removed
+			// the ResourceBinding which dependent on resource template.
+			// So, just return without retry(requeue) would save unnecessary loop.
+			return nil
+		}
+		klog.Errorf("Failed to fetch resource(%s/%s) for resourceBinding. Error: %v.", resource.GetNamespace(), resource.GetName(), err)
+		return err
+	}
+
 	newObj, err := interpreter.AggregateStatus(resource, bindingStatus.AggregatedStatus)
 	if err != nil {
-		klog.Errorf("Failed to aggregate status for resource(%s/%s/%s, Error: %v", gvr, resource.GetNamespace(), resource.GetName(), err)
+		klog.Errorf("Failed to aggregate status for resource(%s/%s/%s), Error: %v", gvr, resource.GetNamespace(), resource.GetName(), err)
 		return err
 	}
 
-	oldStatus, _, _ := unstructured.NestedFieldNoCopy(resource.Object, "status")
-	newStatus, _, _ := unstructured.NestedFieldNoCopy(newObj.Object, "status")
-	if reflect.DeepEqual(oldStatus, newStatus) {
+	if reflect.DeepEqual(resource, newObj) {
 		klog.V(3).Infof("Ignore update resource(%s/%s/%s) status as up to date.", gvr, resource.GetNamespace(), resource.GetName())
 		return nil
 	}
 
-	patchBytes, err := helper.GenReplaceFieldJSONPatch("/status", oldStatus, newStatus)
-	if err != nil {
-		klog.Errorf("Failed to gen patch bytes for resource(%s/%s/%s, Error: %v", gvr, resource.GetNamespace(), resource.GetName(), err)
-		return err
-	}
-
-	_, err = dynamicClient.Resource(gvr).Namespace(resource.GetNamespace()).
-		Patch(ctx, resource.GetName(), types.JSONPatchType, patchBytes, metav1.PatchOptions{}, "status")
-	if err != nil {
+	if _, err = dynamicClient.Resource(gvr).Namespace(resource.GetNamespace()).UpdateStatus(ctx, newObj, metav1.UpdateOptions{}); err != nil {
```

Comment:
use `retry.RetryOnConflict`?

Reply:
hi @whitewindmills. do you mean you prefer following code?

```go
var resource *unstructured.Unstructured
if err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
	// get resource template directly from apiserver, rather than informer cache, detail refers to:
	// https://github.com/karmada-io/karmada/issues/5285
	resource, err = dynamicClient.Resource(gvr).Namespace(objRef.Namespace).Get(ctx, objRef.Name, metav1.GetOptions{})
	if err != nil {
		if apierrors.IsNotFound(err) {
			// It might happen when the resource template has been removed but the garbage collector hasn't removed
			// the ResourceBinding which dependent on resource template.
			// So, just return without retry(requeue) would save unnecessary loop.
			return nil
		}
		klog.Errorf("Failed to fetch resource(%s/%s) for resourceBinding. Error: %v.", resource.GetNamespace(), resource.GetName(), err)
		return err
	}

	newObj, err := interpreter.AggregateStatus(resource, bindingStatus.AggregatedStatus)
	if err != nil {
		klog.Errorf("Failed to aggregate status for resource(%s/%s/%s), Error: %v", gvr, resource.GetNamespace(), resource.GetName(), err)
		return err
	}

	if reflect.DeepEqual(resource, newObj) {
		klog.V(3).Infof("Ignore update resource(%s/%s/%s) status as up to date.", gvr, resource.GetNamespace(), resource.GetName())
		return nil
	}

	if _, err = dynamicClient.Resource(gvr).Namespace(resource.GetNamespace()).UpdateStatus(ctx, newObj, metav1.UpdateOptions{}); err != nil {
		klog.Errorf("Failed to update resource(%s/%s/%s), Error: %v", gvr, resource.GetNamespace(), resource.GetName(), err)
		return err
	}
	return nil
}); err != nil {
	eventRecorder.Event(resource, corev1.EventTypeWarning, events.EventReasonAggregateStatusFailed, err.Error())
	return err
}

eventRecorder.Event(resource, corev1.EventTypeNormal, events.EventReasonAggregateStatusSucceed, "Update Resource with AggregatedStatus successfully.")
klog.V(3).Infof("Update resource(%s/%s/%s) status successfully.", gvr, resource.GetNamespace(), resource.GetName())
```

Reply:
not exactly. `retry.RetryOnConflict` should start after `interpreter.AggregateStatus` because `interpreter.AggregateStatus` may be very time consuming, which causes `resource` to expire.

Reply:
it is recommended to use `helper.UpdateStatus` to update the status.
https://github.com/karmada-io/karmada/blob/5a10d758280c51a8ffe7a279816b08cbd0df14cb/pkg/util/helper/workstatus.go#L78-L85

Reply:
Supposing I have a CRD `XxxDeployment`, its status like: {"a": 1, "b": 1}

filed `b` is controller by resource interpreter `AggregatedStatus`, and is changed from 1 to 2: {"a": 1, "b": 2}

when I run to `UpdateStatus` function, here happens:

field `a` is controlled by my customized controller, and is change from 1 to 2, {"a": 2, "b": 1}

here conflicts occurs, the `XxxDeployment` is out-dated, I should get latest `XxxDeployment` and re-run `AggregatedStatus`.

What do you think of this thing, since `AggregateStatus(oldResource, aggregatedStatusItems)` is deifned by users, we cannot assume that all types of resources are completely covered by latest binding.

Reply:
> field a is controlled by my customized controller, and is change from 1 to 2, {"a": 2, "b": 1}

I don't understand why this scenario exists, and I have never seen such usage. since the user chooses to use the `AggregatedStatus`, the control of the status should be fully handed over to `AggregatedStatus`. so I think your worries are unnecessary.

Reply:
What I want to express is, resource interpreter `AggregateStatus(oldResource, aggregatedStatusItems)` is controlled by two args `oldResource` and `aggregatedStatusItems`, when conflicts, it means `oldResource` is updated, then the output of `AggregateStatus` may be different.

According to your statement, it seems to be assuming that output depends entirely on the `aggregatedStatusItems`, then this resource interpreter shouldn't have two args.

Supposing I have a CRD `XxxDeployment`, its status like: {"a": 1, "b": 1}, however, the field `status.a`  is rely on `spec.c` field, while `spec.c` field is updated by users...

Reply:
lol, so vexing. anyway, I'm sold.

> interpreter.AggregateStatus may be very time consuming, which causes resource to expire.

do you have any idea to resolve it?

Reply:
/cc hi @XiShanYongYe-Chang, JunHua has the same point, what do you think now?

Reply:
I have been persuaded that it is better not to use retry.

Reply:
hi, now, we get resource directly from apiserver and normally only this controller would change its status, so almost impossible conflict. even if there is a conflict, we can perform the reconcile again, no need to use `RetryOnConflict`.

Reply:
> and normally only this controller would change its status

I agree.

> so almost impossible conflict

you cannot guarantee that user controllers won't frequently change their resources, maybe `spec` or `metadata` things. anyway, we can make the update as efficient as possible with `retry.RetryOnConflict` .

> do we still need to stick on this comment?

no, it's up to you. even if you don't take my advice, it's acceptable to me.

Reply:
I got another point, that is, we can't re-use `operationResult, err = UpdateStatus(ctx, c, binding, func() error` such code, since the inner client is `client.Client`, which re-fetch resource from cache, we should use `dynamicClient`.

using `retry.RetryOnConflict` is ok, but since there are no reusable `UpdateStatus` functions, the final code style may still be like this: https://github.com/karmada-io/karmada/pull/5318#discussion_r1709326892, then so that we don't have to trapped in whether need call `interpreter.AggregateStatus` once again.